src/terminal/parser/stateMachine.cpp

```diff
@@ -1270,6 +1437,233 @@ void StateMachine::_EventVt52Param(const wchar_t wch)
+{
+    _trace.TraceOnEvent(L"DcsIntermediate");
+    if (_isStringTerminator(wch))
+    {
+        _EnterGround();
+    }
+    if (_isIntermediate(wch))
+    {
+        _ActionCollect(wch);
+    }
+    else if (_isDcsPassThroughValid(wch))
+    {
+        _ActionDcsPassThrough(wch);
+        _EnterDcsPassThrough();
+    }
+    else if (_isDcsTerminationInitiator(wch))
+    {
+        _EnterDcsTermination();
+    }
+    else
+    {
+        _ActionIgnore();
+    }
+}
+
+// Routine Description:
+// - Processes a character event into an Action that occurs while in the DcsParam state.
+//   Events in this state will:
+//   1. Enter ground on a String terminator
+//   2. Collect DCS parameter data
+//   3. Enter DcsIntermediate if we see an intermediate
+//   4. Enter DcsPassThrough if we see valid DCS characters
+//   5. If we see a ESC, enter the DcsTermination state.
+//   6. Ignore everything else.
+// Arguments:
+// - wch - Character that triggered the event
+// Return Value:
+// - <none>
+void StateMachine::_EventDcsParam(const wchar_t wch)
+{
+    _trace.TraceOnEvent(L"DcsParam");
+    if (_isStringTerminator(wch))
+    {
+        _EnterGround();
+    }
+    else if (_isNumericParamValue(wch) || _isCsiDelimiter(wch))
+    {
+        _ActionParam(wch);
+    }
+    else if (_isIntermediate(wch))
+    {
+        _ActionCollect(wch);
+        _EnterDcsIntermediate();
+    }
+    else if (_isDcsPassThroughValid(wch))
+    {
+        _ActionDcsPassThrough(wch);
+        _EnterDcsPassThrough();
+    }
+    else if (_isDcsTerminationInitiator(wch))
+    {
+        _EnterDcsTermination();
+    }
+    else
+    {
+        _ActionIgnore();
+    }
+}
+
+// Routine Description:
+// - Processes a character event into an Action that occurs while in the DcsPassThrough state.
+//   Events in this state will:
+//   1. Enter ground on a String terminator
+//   2. Pass through if character is valid.
+//   3. If we see a ESC, enter the DcsTermination state.
+//   4. Ignore everything else.
+// Arguments:
+// - wch - Character that triggered the event
+// Return Value:
+// - <none>
+void StateMachine::_EventDcsPassThrough(const wchar_t wch)
+{
+    _trace.TraceOnEvent(L"DcsPassThrough");
+    if (_isStringTerminator(wch))
+    {
+        _EnterGround();
+    }
+    if (_isDcsPassThroughValid(wch))
+    {
+        _ActionDcsPassThrough(wch);
+    }
+    else if (_isDcsTerminationInitiator(wch))
+    {
+        _EnterDcsTermination();
+    }
+    else
+    {
+        _ActionIgnore();
+    }
+}
+
+// Routine Description:
+// - Handle the two-character termination of a DCS sequence.
+//   Events in this state will:
+//   1. Enter ground.
+// Arguments:
+// - wch - Character that triggered the event
+// Return Value:
+// - <none>
+void StateMachine::_EventDcsTermination(const wchar_t wch)
+{
+    _trace.TraceOnEvent(L"DcsTermination");
+
+    if (wch == L'\\')
+    {
+        _EnterGround();
+    }
+
+    // TODO: nested-escape-sequence ST handling

```

Comment:
The idea is from code I found in iTerm2 here:

https://github.com/gnachman/iTerm2/blob/c934e493a0745f68722f959818226d4c98022090/sources/VT100DCSParser.m#L355

Reply:
I think what we want here is something like this:

    if (wch == L'\\')
    {
        // TODO: This is where we'd be dispatching the command.
        _EnterGround();
    }
    else
    {
        _EventEscape(wch);
    }

So if it's not the string terminator, we just treat it as the start of a regular escape sequence, and pass on the character to the escape event.